### ISSUE TITLE
Saving Customer Model directly causes loss of data

### ISSUE THREAD

**Issue by Danielc3** [magento/magento2#12479]
When saving a customer directly using the Customer Model custom attribute data is lost. This occurs when EAV caching is disabled. It appears you can get around this issue by using the Customer Repository or updateData function of the Customer Model. However, even though save is deprecated I would still expect it to work and looking into this has raised some concerns.

What I have found is that when saving directly some attribute set information is missing on the model, if we have a look at the aforementioned updateData function or the Customer Repository save function we see an acknowledgment of this issue in the form of a comment:

` // Need to use attribute set or future updates can cause data loss`

Below this comment both of these functions manually add an attribute set ID based on a hard-coded interface constant. I see no reason why this should not be done in the constructor, as that would get around this issue.

The saving process eventually leads to the checking of whether given attributes are in the expected attribute set using \Magento\Eav\Model\Entity\Attribute\AbstractAttribute::isInSet. What gets passed in this circumstance is null, as the attribute set for the Customer Model is not set. This results in the function returning false and then the attribute value gets deleted.

With regards to why this only occurs with EAV caching disabled. When it is enabled attribute_set_info is not defined on the attributes (what happens is the attribute is cached and then attribute_set_info is added manually afterwards) which is another issue - and one I am yet to fully understand. With attribute_set_info not set isInSet actually returns true and therefore the attribute value is not cleared down.:

```
        if (!$this->hasAttributeSetInfo()) {
            return true;
        }
```

### Preconditions
1. Magento 2.2.0 or 2.1.7

### Steps to reproduce
1. Disable EAV cache.
2. Create a custom attribute for Customer entity.
3. Update a customer in the admin and add a value for this new attribute.
4) Programmatically load the customer, make some change (e.g. firstname) and then save directly using the model.

### Expected result
1. When viewing the customer in the admin we will see the updated firstname and our previously saved custom attribute value.

### Actual result
1. The custom attribute value is lost.

**Comment by magento-engcom-team** [comment]
@Danielc3, thank you for your report.
We've created internal ticket(s) MAGETWO-84716 to track progress on the issue.

**Comment by callumstar** [comment]
Is there any update on the progression of fixing this issue?

**Comment by Vigneshmuthusamy** [comment]
I'm facing the issue with EAV cache enabled as well as disabled in Magento 2.2.2. More than 4 months and still there is no fix for this.

**Comment by Thundar** [comment]
As a workaround, you can create an observer on `customer_save_before` event, with code:

```
        $customerModel = $observer->getData('data_object');
        if (!$customerModel->getAttributeSetId()) {
            $customerModel->setAttributeSetId(
                 \Magento\Customer\Api\CustomerMetadataInterface::ATTRIBUTE_SET_ID_CUSTOMER
             );
         }
```

**Comment by Arkotech123** [comment]
I have same issue, when i try to save customer data it reset custom attributes values. Issue is same for both core and custom

**Comment by sidolov** [comment]
Hi @Danielc3. Thank you for your report. 
The issue has been fixed in magento/magento2#17968 by @Thundar in 2.2-develop branch
Related commit(s):
 - [7965c9bfd532889bca0752a9ef318dae43feb26c](https://github.com/magento/magento2/commit/7965c9bfd532889bca0752a9ef318dae43feb26c)
 - [44fe03f6f067d4f93819b4353f82e2b8caec6c8f](https://github.com/magento/magento2/commit/44fe03f6f067d4f93819b4353f82e2b8caec6c8f)
 - [6a098a7d21b101d1d1ee058917a15ff6ea4406b4](https://github.com/magento/magento2/commit/6a098a7d21b101d1d1ee058917a15ff6ea4406b4)

The fix will be available with the upcoming 2.2.8 release.

**Comment by slavvka** [comment]
Hi @Danielc3. Thank you for your report. 
The issue has been fixed in magento/magento2#18571 by @gelanivishal in 2.3-develop branch
Related commit(s):
 - [2322baf10841259b122dba4350f7c2cbcbe2bba3](https://github.com/magento/magento2/commit/2322baf10841259b122dba4350f7c2cbcbe2bba3)
 - [ce82c5ce4784125429b0baa10478277dda3d689c](https://github.com/magento/magento2/commit/ce82c5ce4784125429b0baa10478277dda3d689c)
 - [97076762561f7781ab2a8e36e6a7d29556d5e98e](https://github.com/magento/magento2/commit/97076762561f7781ab2a8e36e6a7d29556d5e98e)

The fix will be available with the upcoming 2.3.1 release.
